src/CPULSSolver.cpp

```diff
@@ -313,28 +298,41 @@ void CPULSSolver::computeFSRSources(int iteration) {
         src_y = scatter_source_y + chi[g] * fission_source_y;
         src_z = scatter_source_z + chi[g] * fission_source_z;
 
-        /* Compute total (scatter+fission+fixed) reduced source moments */
+        /* Compute total (scatter+fission) reduced source moments */
         if (_solve_3D) {
-          _reduced_sources_xyz(r,g,0) = ONE_OVER_FOUR_PI / 2 *
-              (_FSR_lin_exp_matrix[r*num_coeffs  ] * src_x +
-               _FSR_lin_exp_matrix[r*num_coeffs+2] * src_y +
-               _FSR_lin_exp_matrix[r*num_coeffs+3] * src_z);
-          _reduced_sources_xyz(r,g,1) = ONE_OVER_FOUR_PI / 2 *
-              (_FSR_lin_exp_matrix[r*num_coeffs+2] * src_x +
-               _FSR_lin_exp_matrix[r*num_coeffs+1] * src_y +
-               _FSR_lin_exp_matrix[r*num_coeffs+4] * src_z);
-          _reduced_sources_xyz(r,g,2) = ONE_OVER_FOUR_PI / 2 *
-              (_FSR_lin_exp_matrix[r*num_coeffs+3] * src_x +
-               _FSR_lin_exp_matrix[r*num_coeffs+4] * src_y +
-               _FSR_lin_exp_matrix[r*num_coeffs+5] * src_z);
+          if (_reduced_sources(r,g) > 1e-15 or iteration > 29) {
```

Comment:
It seems here if the flat source component is zero, you force the gradient of the source to be zero. It is supposed to be so, but do you have any theory bases?
Also, in Geoff's thesis page 155 the iteration number is 20 rather than 30 here. @GiudGiud